src/rm_cms.c

```diff
@@ -0,0 +1,309 @@
+#include <strings.h>        // strncasecmp
+#include <stdlib.h>         // malloc
+#include <math.h>           // ceil, log10f
+
+#include "version.h"
+#include "rmutil/util.h"
+
+#include "rm_cms.h"
+#include "cms.h"
+
+RedisModuleType *CMSketchType;
+
+typedef struct {
+    const char *key;
+    long long value;
+} CMSPair;
+
+static int CompareCStringToRedisString(char *str, RedisModuleString *redisStr) {
+    size_t len;
+    char *cRedisStr = (char *)RedisModule_StringPtrLen(redisStr, &len);
+    return strncasecmp(cRedisStr, str, len) == 0;
+}
+
+static int GetCMS(RedisModuleCtx *ctx, RedisModuleString *keyName, 
+                    CMSketch **cms, int mode) {
+    RedisModuleKey *key = RedisModule_OpenKey(ctx, keyName, mode);
+    if(RedisModule_KeyType(key) == REDISMODULE_KEYTYPE_EMPTY) {
+        RedisModule_ReplyWithError(ctx, "CMS: key does not exist");
+        return REDISMODULE_ERR;
+    } else if(RedisModule_ModuleTypeGetType(key) != CMSketchType) {
+        RedisModule_ReplyWithError(ctx, REDISMODULE_ERRORMSG_WRONGTYPE);
+        return REDISMODULE_ERR;
+    } 
+    *cms = RedisModule_ModuleTypeGetValue(key);
+    return REDISMODULE_OK;
+}
+
+static int CreateCmsKey(RedisModuleCtx *ctx, RedisModuleString *keyName, 
+        long long width, long long depth, CMSketch **cms, RedisModuleKey **key) {
+    if(*key == NULL) {
+        *key = RedisModule_OpenKey(ctx, keyName, REDISMODULE_READ|REDISMODULE_WRITE);
+    }
+
+    RedisModule_RetainString(ctx, keyName);
```

Comment:
Why is that needed? And if it is, where is the free?

Reply:
Not sure :(
It was referenced as black magic. 
Will study further.

Reply:
Always keep learning.

Reply:
Retain/Free work as reference counter increments/decrements

Reply:
Removed